Add basic support for resource quota

atantawi opened the pull request:
Handles phase 1 of issue #44, namely Basic handling of quota resources {cpu, memory, ...}, {requests.cpu, requests.memory, ...} and {limits.cpu, limits.memory, ...}, including extended resources.

dgrove-oss commented on deployment/mcad-controller/templates/clusterrole.yaml:
```diff
@@ -218,3 +218,11 @@ rules:
   verbs:
   - create
+- apiGroups:
+  - ""
```
```suggestion
  -  workload.codeflare.dev
```

dgrove-oss replied:
I think it should match the change to config/rbac/role.yaml,

dgrove-oss replied:
I got it backwards; will update

dgrove-oss commented on config/rbac/role.yaml:
```diff
@@ -220,3 +220,17 @@ rules:
   - patch
   - update
+- apiGroups:
+  - workload.codeflare.dev
```
```suggestion
  - ""
```

atantawi replied:
Done

dgrove-oss commented on internal/controller/appwrapper_controller.go:
```diff
@@ -34,6 +34,8 @@ import (
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=resourcequotas,verbs=get;list;watch
```
```suggestion
//+kubebuilder:rbac:groups="",resources=resourcequotas,verbs=get;list;watch
```

atantawi replied:
Done

dgrove-oss commented on internal/controller/appwrapper_controller.go:
```diff
@@ -34,6 +34,8 @@ import (
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=resourcequotas,verbs=get;list;watch
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=resourcequotas/status,verbs=get
```
```suggestion
//+kubebuilder:rbac:groups="",resources=resourcequotas/status,verbs=get
```

atantawi replied:
Done

atantawi commented:
Fixed kubebuilder directive in `internal/controller/appwrapper_controller.go` so that `config/rbac/role.yaml` and `deployment/mcad-controller/templates/rbac/clusterrole.yaml` are matched.
(Also, rebased on main to pick up commit #141.)

Now E2E test fails due to:
`Internal error occurred: failed calling webhook "webhook.cert-manager.io": failed to call webhook: Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp 10.96.50.154:443: connect: connection refused`

atantawi commented:
> Fixed kubebuilder directive in `internal/controller/appwrapper_controller.go` so that `config/rbac/role.yaml` and `deployment/mcad-controller/templates/rbac/clusterrole.yaml` are matched. (Also, rebased on main to pick up commit #141.)
> 
> Now E2E test fails due to: `Internal error occurred: failed calling webhook "webhook.cert-manager.io": failed to call webhook: Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp 10.96.50.154:443: connect: connection refused`

Fixed with commit #144

tardieu commented on internal/controller/dispatcher.go:
```diff
@@ -257,3 +265,41 @@ func (r *Dispatcher) dispatch(ctx context.Context) (ctrl.Result, error) {
+	for _, appWrapper := range appWrappers.Items {
+		phase, step := r.getCachedAW(&appWrapper)
+		if phase == mcadv1beta1.Running &&
+			step == mcadv1beta1.Idle || step == mcadv1beta1.Creating || step == mcadv1beta1.Created {
```
This condition does not seem right. Just use: `step != mcadv1beta1.Idle`.

atantawi replied:
done

tardieu commented on internal/controller/dispatch_logic.go:
```diff
@@ -237,6 +262,18 @@ func aggregateRequests(appWrapper *mcadv1beta1.AppWrapper) Weights {
+	for _, r := range appWrapper.Spec.Resources.GenericItems {
+		for _, cpr := range r.CustomPodResources {
+			// TODO: check why not implemented limits spec in CustomPodResource?
+			limit.AddProd(cpr.Replicas, NewWeights(cpr.NotImplemented_Limits))
```
Now that limits are relevant, please update the CRD renaming `NotImplemented_Limit` to `Limits`.

atantawi replied:
done

tardieu commented on internal/controller/quota_tracker.go:
```diff
@@ -0,0 +1,149 @@
+		"QuotaWeights", quotaState.quota, "UsedWeights", quotaState.used,
+		"AllocatedWeights", quotaState.allocated, "unAdmittedWeights", unAdmittedWeights,
+		"AvailableWeights", quotaWeights, "appWrapperAskWeights", appWrapperAskWeights,
+		"quotaFits", quotaFits)
```
Please disable this log line by default as it is going to make logging very verbose.

atantawi replied:
done

tardieu commented on internal/controller/dispatcher.go:
```diff
@@ -257,3 +265,41 @@ func (r *Dispatcher) dispatch(ctx context.Context) (ctrl.Result, error) {
+
+// Calculate resource demands of appWrappers that have been dispatched but haven't
+// passed through ResourceQuota admission controller yet (approximated by resources not created yet)
+func (r *Dispatcher) getUnadmittedAppWrappersWeights(ctx context.Context) (map[string]*WeightsPair, error) {
```
Maybe rename to `getUnadmittedAppWrapperPodsWeights`?

atantawi replied:
done

tardieu commented on internal/controller/dispatch_logic.go:
```diff
@@ -204,13 +204,38 @@ func (r *Dispatcher) selectForDispatch(ctx context.Context) ([]*mcadv1beta1.AppW
+		insufficientResources := []v1.ResourceName{}
+		if len(resourceQuotas.Items) > 0 {
+			appWrapperAskWeights = getWeightsPairForAppWrapper(appWrapper)
+			// TODO: relax assumption that only one resourceQuota per nameSpace
```
Please open an issue (or comment on an existing issue) to document this limitation and others.

atantawi replied:
added to issue #44 .

tardieu reviewed (approved):
lgtm
